### PR TITLE
tools/bootloader: add force-erase option and add bootloader version (Sponsored by CubePilot)

### PR DESCRIPTION
**Pull request by julianoes**
This PR does two things:

## Add force-erase option

If the STM32H7 fails to program or erase a full chunk of 256 bytes, the ECC check will trigger a busfault when trying to read from it.

To speed up erasing and optimize wear, we read before erasing to check if it actually needs erasing. That's when a busfault happens and the erase time outs.

The workaround is to add an option to do a full erase without check.

Credit goes to:
https://github.com/ArduPilot/ardupilot/pull/22090

And the protocol option added to the bootloader is the same as for ArduPilot, so compatible.

## Add bootloader version

This PR also adds a new **bootloader version** essentially a string that we can set in the format of: `PX4BLv1.15.0g12345678` constrained to max 32 chars. This allows to check if a bootloader needs updating while not breaking any tooling by bumping the **bootloader protocol revision** which has been stabilized at 5 for a long time by now.

The output of the px_uploader.py now looks as follows:

```
uploading px4
Waiting for bootloader...

Found board id: 140,0 bootloader protocol revision 5 on /dev/serial/by-id/usb-CubePilot_PX4_BL_CubePilot_CubeOrange_0-if00
Loaded firmware for board id: 140,0 size: 1902144 bytes (96.75%) 

Bootloader version: PX4BLv1.15.0g3bbeb0d87d
Sn: 003100433230510b39343831
Chip: 20036450
Family: STM32H7[4|5]x
Revision: V
Flash: 1966080 bytes
Windowed mode: False

Erase  : [====================] 100.0%
Program: [======              ] 33.9%
```

### CODE REVIEW

File: Tools/px_uploader.py
Comment (davids5) [inline Tools/px_uploader.py:801]:
```suggestion
    parser.add_argument('--force-erase', action="store_true", help="Do not perform the blank check, always erase every sector of the application space")
```

**Review by davids5 — CHANGES_REQUESTED** [review]
@julianoes Nice! The wording on the option felt wrong to me. See if what I added works for you.

**Comment by julianoes** [comment]
Fixed, thanks @davids5!

**Review by davids5 — CHANGES_REQUESTED** [review]
Thinking about this a bit more....Since this is s protocol change should we bump the [rev](https://github.com/PX4/PX4-Autopilot/pull/22777/files#diff-18abc5087507677cf970069a8a5b3c8a2f9886ce00cc21a347e33af9312e83a8L83) and add it to https://github.com/PX4/PX4-Bootloader?

**Comment by julianoes** [comment]
@davids5 we should *but* we haven't in the past, and it looks like ArduPilot has many more features and is also still on the same revision.

**Comment by julianoes** [comment]
@davids5 it doesn't need to go into the Bootloader repo, because it only applies for H7, I think.

**Comment by julianoes** [comment]
@davids5 I had messed up. Here is a fixup commit: [ac5b06f](https://github.com/PX4/PX4-Autopilot/pull/22777/commits/ac5b06fa0fe2c6df99f3e3de32061c23b3c0e29d) It would be great if you could double check that line again. Thanks!

File: platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c
Comment (davids5) [reply platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:412]: Let's convert this to  an int return (like up_progmem_ispageerased(sector) != 0) an use the != 0 for Not blank

File: platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c
Comment (davids5) [inline platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:431]: so this reads `if (force || flash_func_is_sector_blank(sector) != 0)`

**Review by davids5 — CHANGES_REQUESTED** [review]
See inlines

File: platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c
Comment (julianoes) [reply platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:412]: Ok!

File: platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c
Comment (julianoes) [reply platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:431]: Ok!

**Comment by julianoes** [comment]
Pushed another fixup @davids5.

File: platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c
Comment (davids5) [inline platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:404]:
```suggestion
	int blank = 0; /* Assume it is Bank */
```

File: platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c
Comment (davids5) [inline platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:413]:
```suggestion
	return blank;
```

File: platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c
Comment (davids5) [inline platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:398]:
```suggestion
/* imxRT uses Flash lib, not up_progmem so let's stub it here */
up_progmem_ispageerased(unsigned sector)
```

**Review by davids5 — CHANGES_REQUESTED** [review]
I thing it would be good to use 
up_progmem_ispageerased as the name, see comment

**Comment by julianoes** [comment]
@davids5 ok, I've accepted your suggestions. I'll run CI, then rebase, squash and force push.

**Review by davids5 — COMMENTED** [review]
On the issue of the revision - The uploader side can blindly send the command and it can fail silently.  

As far as putting this in the the bootloader repo: It would be a nice to have but will not fit on all the SoC.

File: platforms/nuttx/src/bootloader/common/bl.c
Comment (davids5) [inline platforms/nuttx/src/bootloader/common/bl.c:653]:
```suggestion
		bool force_erase = false;
```

**Review by davids5 — COMMENTED** [review]
The rest all looks good. This is just a nit.  I will leave it up to you, if you want to push again :)

**Comment by julianoes** [comment]
@davids5 on second thought, I think you're right that it would be worthwhile to bump the version number.

**Review by davids5 — DISMISSED** [review]
Ok Looks good!

**Comment by julianoes** [comment]
@dagar what's an easy way to update all bootloader images? Do I have to do it one by one?

**Comment by davids5** [comment]
@julianoes There is a target  iirc. something like   update_bootloaders

**Comment by julianoes** [comment]
@davids5 thanks, I'll try that. However, I wonder whether we should really bump the protocol version. The problem is that QGC refused to flash bootloader rev 6, and so does the uploader script, and that's even though they would just work fine. Instead of a "major" change it's more like a hidden extra feature. On the other hand, it's hard to know whether a flashed bootloader supports the force-erase option. You basically don't know until you try.

**Comment by davids5** [comment]
> @davids5 thanks, I'll try that. However, I wonder whether we should really bump the protocol version. The problem is that QGC refused to flash bootloader rev 6, and so does the uploader script, and that's even though they would just work fine. Instead of a "major" change it's more like a hidden extra feature. On the other hand, it's hard to know whether a flashed bootloader supports the force-erase option. You basically don't know until you try.

So it would force an update for QGC and the uploader on the users. How bad is that?

**Comment by julianoes** [comment]
> How bad is that?

Not the end of the world but still unexpected.

**Comment by julianoes** [comment]
@davids5 this has evolved a bit. @bugobliterator and I came up with a **bootloader version** additional to the **bootloader protocol revision**. For more info, see the PR description and the individual commits. I also took the liberty to clean up the Python script a bit.

I'd appreciate your review @davids5.

FYI @dagar

**Comment by davids5** [comment]
> @davids5 this has evolved a bit. @bugobliterator and I came up with a **bootloader version** additional to the **bootloader protocol revision**. For more info, see the PR description and the individual commits. I also took the liberty to clean up the Python script a bit.
> 
> I'd appreciate your review @davids5.
> 
> FYI @dagar
> 
> TODO: re-generate all bootloaders from here.

@julianoes - can we have a call on this? I am in US EST so maybe 4 AM my time next week or 03:00 PM EST. Let me know

**Comment by julianoes** [comment]
@davids5 I have the feeling the GET_SUPPORTED_COMMANDS API that you suggested is - unfortunately - not really worth the trouble.

It seems worthwhile because it would solve the problem where we don't know if force-erase is supported or not, basically if GET_SUPPORTED_COMMANDS is not supported, force-erase won't be either. However, this doesn't hold for the ArduPilot bootloader, so we're sort of back at square one.

I agree that it would only be worthwhile for the next bootloader addition down the road, but it doesn't seem quite worth it, because you can easily just "try" every command, except erase which take a long time to time out.

File: Tools/px_uploader.py
Comment (davids5) [inline Tools/px_uploader.py:467]: Wouldn't the the force erase bet met with an immediate command error? In that case should you attempt to use an erase or suggest a bootloader upgrade?

**Review by davids5 — CHANGES_REQUESTED** [review]
See inline.

File: Tools/px_uploader.py
Comment (julianoes) [reply Tools/px_uploader.py:467]: Yes, that would be nice. But a unknown byte seems to just be ignored. And the erase command is not acked until it is erased, so you have to wait a long time. Not optimal I know...

Comment (davids5) [reply Tools/px_uploader.py:467]: geez That seems wrong, bit it is what the code does.

**Review by davids5 — DISMISSED** [review]

File: Tools/px_uploader.py
Comment (julianoes) [reply Tools/px_uploader.py:467]: Yep.

**Comment by julianoes** [comment]
Final rebase, bootloader rebuild and push. As soon as CI is happy I'll merge.

**Comment by julianoes** [comment]
CI results are the same as main. Merging.

**Comment by PetervdPerk-NXP** [comment]
FMU-V6XRT: Build regression on main since this PR.
```
/home/peter/src/Firmware/platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:399:1: error: return type defaults to 'int' [-Werror=implicit-int]
  399 | up_progmem_ispageerased(unsigned sector)
      | ^~~~~~~~~~~~~~~~~~~~~~~
compilation terminated due to -Wfatal-errors.
cc1: all warnings being treated as errors
```

**Comment by davids5** [comment]
> `px4_fmu-v6xrt`: Build regression on main since this PR.
> 
> ```
> /home/peter/src/Firmware/platforms/nuttx/src/bootloader/nxp/imxrt_common/main.c:399:1: error: return type defaults to 'int' [-Werror=implicit-int]
>   399 | up_progmem_ispageerased(unsigned sector)
>       | ^~~~~~~~~~~~~~~~~~~~~~~
> compilation terminated due to -Wfatal-errors.
> cc1: all warnings being treated as errors
> ```
> 
> Commit [4a13e49](https://github.com/PX4/PX4-Autopilot/commit/4a13e495d7b7dfc743314009e3a22d828260f49c) was succesful on `px4_fmu-v6xrt` before this PR.

Fixed in https://github.com/PX4/PX4-Autopilot/pull/23174

**Pull request merged**
